### PR TITLE
Implemented overrideToFallbackLanguage flag and fallbackLanguage;

### PR DESCRIPTION
**Pull request by Antonijo2307**
Add functionality that enables user to configure Swiftl18n to use fallback language in case there is no translation for searched key.

Productive task:
https://app.productive.io/1-infinum/task/8092739

### CODE REVIEW

**Comment by Antonijo2307** [comment]
The problem I see is that there is no way to determine for certain that there is no translation for a specific key, there are no constraint on key naming so it can easily be called the same as translation. I don't know if you all had some idea on how to resolve this issue, cause our translation tables after they are generated by polyglot already have filled empty translations with key values. 
Maybe we should leave those values empty and if empty return key values or fallback value? Or constraint key naming to some recognisable format like key_some_key_neme... What do you think?

File: SwiftI18n/Classes/Main/I18nManager.swift [hunk 1]
```diff
@@ -12,13 +12,27 @@ public class I18nManager {
...
+    public var fallbackLanguage: String?
+    
     private init() {}
     
     public var defaultLanguage: String?
     public var availableLanguages: [String]?
     
-    public var localizationPerformingBlock: (_ key: String, _ language: String)->(String) = { (key, language) in
-        return NSLocalizedString(key, tableName: language, comment: "")
+    public lazy var localizationPerformingBlock: (_ key: String, _ language: String)->(String) = { [weak self] (key, language) in
```
Comment (Truba) [inline SwiftI18n/Classes/Main/I18nManager.swift:23]: Hi @Antonijo2307 I don't think this is a good approach since it introduces `major chages` to the library.
`localizationPerformingBlock` is public and was designed to allow anyone to add set their own implementation.
--
In general what you are doing here can already be done in your project, no need to change this library.
--
If you want to, you can add it as a convenience  function that will set `localizationPerformingBlock` to what you need.

Something like:
```
public static func localizationPerformingBlockWithFallbackLanguage(fallbackLanguage: String) = { [weak self] (key, language) in
...
}
```
and then in your code you can simply do:
`Manager.localizationPerformingBlock = Manager.localizationPerformingBlockWithFallbackLanguage('your_language')`

File: SwiftI18n/Classes/Main/I18nManager.swift [hunk 2]
```diff
@@ -12,13 +12,27 @@ public class I18nManager {
 
     public static var instance = I18nManager()
     
+    public var overrideToFallbackLanguage = false
+    public var fallbackLanguage: String?
```
Comment (Truba) [inline SwiftI18n/Classes/Main/I18nManager.swift:16]: Also this is not idiomatic Swift.
You should only have:
`public var fallbackLanguage: String?` - if it's null, then do nothing, if it's some then override.

With this you can have: overrideToFallbackLanguage: true & fallbackLanguage:nil | overrideToFallbackLanguage: false & fallbackLanguage:some which is not a valid state.

File: SwiftI18n/Classes/Main/I18nManager.swift (hunk 1 above)
Comment (Antonijo2307) [reply SwiftI18n/Classes/Main/I18nManager.swift:23]: It was task requirement to add it to I18nManager.instance, I thought you all agreed on it. If this is not necessary just let me know and I will close this PR...

File: SwiftI18n/Classes/Main/I18nManager.swift (hunk 2 above)
Comment (Antonijo2307) [reply SwiftI18n/Classes/Main/I18nManager.swift:16]: I agree, my thought exactly, again task said explicitly to use these 2 members, my bad, should have optimised it during implementation

Comment (jabou) [reply SwiftI18n/Classes/Main/I18nManager.swift:16]: In task, we didn't define that fallbackLanguage is optional. If we set it as optional value, then we can go only with one property and then we don' need bool property.

The main idea is to give possibility to user to fallback in production, while keeping keys (instead of missing translaiton) on staging builds so QA can easier check where we miss translation. E.g.:
```swift
#if STAGING
I18nManager.instance.fallbackLanguage = nil
#elseif APPSTORE
I18nManager.instance.fallbackLanguage = Language.eng.code
#endif
```

File: SwiftI18n/Classes/Main/I18nManager.swift (hunk 1 above)
Comment (jabou) [reply SwiftI18n/Classes/Main/I18nManager.swift:23]: Hm okay, let me go through this comment:
> In general what you are doing here can already be done in your project
The main problem is that no one is aware of this, and that's why we got the request from some projects. The issue is that the client doesn't want to see the translation key in places where the translation is missing for the production targets.
I'm okay with your proposal to create a new block, `localizationPerformingBlockWithFallbackLanguage,` which will generate practically the same outcome. Replace missing translations with some language (usually the default one).

> to not pollute the library with project specific things
It was mentioned in a few projects, so it's more than one specific project. We want to push our teams to start doing this: showing the key on the screen on staging builds and showing the default language translation on production.

> If you need to, you can add it as a convenience function that will set localizationPerformingBlock to what you need.
> ...
> and then in your code you can simply do
If the output is the same, and it's easy for the lib user as you defined it here, then I'm okay with that approach! Let's do it that way, then, without the proposed properties.

Comment (Truba) [reply SwiftI18n/Classes/Main/I18nManager.swift:23]: > We want to push our teams to start doing this: showing the key on the screen on staging builds and showing the default language translation on production.

Hey @jabou , I don't think this is a good idea. It seems you are trying to solve a wrong problem.
Having a fallback translation instead of the correct one is still a bug.

The problem you have is that not all the translation are filled in.
The way to solve it is to make sure that all the translations are filled in :) 

This is something that should be addressed, for example: "Check all translations are added" button can be added to polyglot web, and your project's release process should include someone pressing that button.

**Comment by Truba** [comment]
@jabou
 @Antonijo2307 also had a good comment. Take for example:
translationKey: "ok"
english translation: "ok"
croatian translation: "dobro"
If user set's english as preferred language, and croatian as a fallback they would get "dobro" as a translation.

File: SwiftI18n/Classes/Main/I18nManager.swift (hunk 1 above)
Comment (jabou) [reply SwiftI18n/Classes/Main/I18nManager.swift:23]: I totally agree with you that this should not end in production, but that's not how some of our clients work. 

To mention a few use cases we are trying to cover with this:
1. Translation is added for a future release and is not currently used in the app. The button will not cover that case, as it's okay not to have the translation at this moment.
2. The client has an external company that adds missing translations. Sometimes, that company deletes a translation or forgets to add it. In that case, rather than showing `some_translation_key` on screen, most apps show the "default" (usually English) translation as a mix of two languages (still not great, but much better than showing just a raw translation key). I once even saw a mix of German and English words in Apple settings (places where German translation was missing). That's something we want to accomplish; if someone breaks something with translation, we want to fall back to the default language on production, while on staging builds, we want to show the translation key for easier debugging. Of course, with the flexibility to always show keys if the client wants or to always show fallback if that's the case they want on staging.

_We will not cover the case if someone deliberately puts the translation key value under the default translation box._

**Comment by jabou** [comment]
>@jabou
@Antonijo2307 also had a good comment. Take for example:
translationKey: "ok"
english translation: "ok"
croatian translation: "dobro"
If user set's english as preferred language, and croatian as a fallback they would get "dobro" as a translation.


Maybe we can check if the translation has `_`? In your case, we can assume that the "ok" key is translated as the value key does not contain "_". 
In the case of checking the `_`, I would add some "smarter" logic than just checking if the value contains `_` as this will not cover the translation, e.g., `"Fill empty lines: _"` (e.g., hint/subtitle). 
We can separate text by `_`, and if there are more than 1 or 2 elements in the array, -> not translated. Or check if we have `_` *and* `" "` (spaces) -> translated. Something like that should cover cases without `_`.

**Comment by Antonijo2307** [comment]
Or we can leave them empty during translation table generation, and then just handle empty string, if empty use key or default translation, depending on the overrideToFallbackLanguage  flag value. It should be easy to implement, older project can choose to stay as they are or just do "polyglot pull" again to clean translation tables.

**Comment by Antonijo2307** [comment]
@jabou && @Truba

**Comment by jabou** [comment]
@Antonijo2307 This will break debugging; the main purpose of showing the key instead of the translation is for QA (and the client) to easily see what translations are missing on the screen. 
However, I'm all ears if you can combine those stuff somehow.

**Comment by Antonijo2307** [comment]
No breaking of debug:

if translation.isEmpty {
    if overrideToFallbackLanguage {
        translation = falbackLanguageTranslation
    } else {
        translation = key_value
    }
}

so it will be the same result as you defined it in this task, if you set overrideToFallbackLanguage = false then you will se keys in places that do not have translations...

File: SwiftI18n/Classes/Main/I18nManager.swift [hunk 3]
```diff
@@ -12,13 +12,27 @@ public class I18nManager {
 
     public static var instance = I18nManager()
     
+    public var overrideToFallbackLanguage = false
```
Comment (jabou) [inline SwiftI18n/Classes/Main/I18nManager.swift:15]: As agreed in the comments, feel free to remove this one and just keep `fallbackLanguage: String?`.

File: SwiftI18n/Classes/Main/I18nManager.swift (hunk 1 above)
Comment (jabou) [reply SwiftI18n/Classes/Main/I18nManager.swift:23]: @Antonijo2307, can you please check those comments and give us your input? Regarding the braking change, I agree with Vlaho that we should not change public methods/properties in this way, as this could potentially break someone's active implementation. 

Please check if you can change how this can be handled from the comments Vlaho left in a thread (e.g. `localizationPerformingBlockWithFallbackLanguage` approach).

**Comment by Antonijo2307** [comment]
@jabou I don't see how it can break someones implementation, if someone defined his own block than this one will just not be called anyways. The solution with predefined localizationPerformingBlockWithFallbackLanguage block does the same thing, if I define block "localizationPerformingBlockWithFallbackLanguage" on manager and I set that one from my project than I can't use my custom block if I used one before or if I use my custom block than I can't use localizationPerformingBlockWithFallbackLanguage block, right?. So it doesn't really solve the problem. As @Truba said, the problem can be resolved locally per project, and anyone who has custom block implemented will have to deal with it by updating code in their custom localisation block, we can make some instructions on how to do it. But I think most users use this predefined localisation block, and if they did not define documentation language it will behave the same as before, so again no broken implementation. 

So I can ditch the lazy part by declaring fallbackLanguage as static var, manager is singletone with private constructor after all, and do something like this:

```
public var localizationPerformingBlock: (_ key: String, _ language: String)->(String) = { (key, language) in

        let translation = NSLocalizedString(key, tableName: language, comment: "")
        
        guard translation == key else {
            return translation
        }

        guard let fallbackLanguage = I18nManager.fallbackLanguage else {
            return translation
        }
        
        return NSLocalizedString(key, tableName: fallbackLanguage, comment: "")
    }
```

What do you think? Do you have some usecase to demonstrate breaking of functionality that I didn't account for?

**Comment by goranbrl** [comment]
@jabou @Truba @Antonijo2307 what's the status of this PR?

**Comment by Antonijo2307** [comment]
@goranbrl We are still trying to work out some implementation details. @jabou said he will take a look at my questions as soon as he gets a chance.

**Comment by jabou** [comment]
@Antonijo2307 I checked the proposed implementation, and it still breaks the logic. As long as `localizationPerformingBlock` is a variable that anyone can change, our logic will stop working.

Have you checked maybe if something like this will work:
Return original logic of `localizationPerformingBlock`:
```swift
    public var localizationPerformingBlock: (_ key: String, _ language: String)->(String) = { (key, language) in
        return NSLocalizedString(key, tableName: language, comment: "")
    }
```

and  then move your logic into the `localizedString(forKey:)`:
```swift
    func localizedString(forKey key: String) -> String {
        let translation = localizationPerformingBlock(key, language)

        guard translation == key else {
            return translation
        }

        guard let fallbackLanguage else {
            return translation
        }

        return localizationPerformingBlock(key, fallbackLanguage)
    }
```

In this case, if the user sets `fallbackLanguage` (you can remove it from `static` -> `public var fallbackLanguage: String?`), the user's logic for `localizationPerformingBlock` will still stay the same, except it will be called twice if we have `fallbackLanguage` and primary language translation is missing. 
If they have some custom logic within `localizationPerformingBlock`, by adding the `fallbackLanguage`, the will still keep the logic that we want to accomplish. 
In your current proposal, if user set the `fallbackLanguage` and implements their own `localizationPerformingBlock`, that property will not do anything and it will look like a bug in the lib.

**Comment by Antonijo2307** [comment]
Ok I have updated the code. Now the block is free for implementation and internally LocText doesn't call it directly. This solves most of the problems from above. 
Problems that are not solved are the ones where key is just one word, like:
ok, done, save...
All these examples will have the same translations in english, and if english is not fallback language, they will be translated
I don't see any solution but to standardise key naming to lets say key_[something]. I don't know if we can do this but that is the only solution I can see. @Truba @jabou what do you think?

**Comment by jabou** [comment]
@Antonijo2307 
> but to standardise key naming to lets say key_[something]

I would agree, but that should not be part of this lib. We can write that limitation when fallback translation is used in the readme (documentation) with the proposed solution.

File: SwiftI18n/Classes/Main/I18nManager.swift [hunk 4]
```diff
@@ -12,6 +12,8 @@ public class I18nManager {
 
     public static var instance = I18nManager()
     
+    public static var fallbackLanguage: String?
+    
     private init() {}
     
     public var defaultLanguage: String?
```
Comment (jabou) [inline SwiftI18n/Classes/Main/I18nManager.swift:19]:
```suggestion
    public static var instance = I18nManager()
    private init() {}
    
    public var defaultLanguage: String?
    public var fallbackLanguage: String?
```

This does not need to be static anymore. Let's remove that so that the call site stays same for the fallback as it is for the default.

File: SwiftI18n/Classes/Main/I18nManager.swift [hunk 5]
```diff
@@ -58,8 +60,20 @@ public extension I18nManager {
 
 public extension I18nManager {
     
-    func localizedString(forKey key: String) -> String {
-        return localizationPerformingBlock(key, language)
+    func localizedString(forKey key: String, language: String? = nil) -> String {
+
+        var localizedString = localizationPerformingBlock(key, language ?? self.language)
```
Comment (jabou) [inline SwiftI18n/Classes/Main/I18nManager.swift:65]:
```suggestion
    func localizedString(forKey key: String, language: String) -> String {

        var localizedString = localizationPerformingBlock(key, language)
```

Why did you change this? It will introduce breaking change that we want to avoid.

File: SwiftI18n/Classes/Main/I18nManager.swift [hunk 6]
```diff
@@ -58,8 +60,20 @@ public extension I18nManager {
...
+        
+        guard localizedString == key else {
+            return localizedString
+        }
+
+        guard let fallbackLanguage = I18nManager.fallbackLanguage else {
```
Comment (jabou) [inline SwiftI18n/Classes/Main/I18nManager.swift:71]:
```suggestion
        guard let fallbackLanguage = fallbackLanguage else {
```

File: SwiftI18n/Classes/Views/SwiftUI/LocText.swift [hunk 7]
```diff
@@ -60,7 +60,7 @@ private extension LocText {
         func text(localizedIn language: String) -> Text {
             switch self {
             case .key(let key):
-                Text(I18nManager.instance.localizationPerformingBlock(key, language))
+                Text(I18nManager.instance.localizedString(forKey: key, language: language))
```
Comment (jabou) [inline SwiftI18n/Classes/Views/SwiftUI/LocText.swift:63]: Do we need to change this? 🤔

Comment (Antonijo2307) [reply SwiftI18n/Classes/Views/SwiftUI/LocText.swift:63]: If we want it to work, than I would say yes

File: SwiftI18n/Classes/Main/I18nManager.swift (hunk 5 above)
Comment (Antonijo2307) [reply SwiftI18n/Classes/Main/I18nManager.swift:65]: Can you explain, I don't see a breaking change? Before there was no language parameter, and it worked with language property which is a member of I18nManager, I don't see a major change here with this optional parameter. 
But if it is a big concern I can implement a different method for this

Comment (Antonijo2307) [reply SwiftI18n/Classes/Main/I18nManager.swift:65]: We reviewed the comments and decided to go wi the changes and add some comments to method declaration

File: SwiftI18n/Classes/Main/I18nManager.swift (hunk 1 above)
Comment (Antonijo2307) [reply SwiftI18n/Classes/Main/I18nManager.swift]: Ok, we decided not to touch the block it self but to implement the changes in method that calls that block.

File: SwiftI18n/Classes/Main/I18nManager.swift (hunk 2 above)
Comment (Antonijo2307) [reply SwiftI18n/Classes/Main/I18nManager.swift]: We are using only one property now

File: SwiftI18n/Classes/Main/I18nManager.swift [hunk 8]
```diff
@@ -17,10 +17,29 @@ public class I18nManager {
...
+    /// ```
+    ///
+    /// **Important Note**: When `fallbackLanguage` is set, the system will attempt to fetch translations
+    /// in this fallback language if the key matches translation in preferred language. This behavior ensures that
+    /// missing localizations are substituted with a consistent backup language.
+    ///
+    /// **Risks**:
+    /// - If a localization key is the same as the translation (e.g., common terms like "ok," "done," or "cancel"),
+    ///   and the `fallbackLanguage` is not set to English, the system will return a translation in the fallback
+    ///   language, even if an appropriate translation exists in the preferred language, this can lead to unintended substitutions
```
Comment (jabou) [inline SwiftI18n/Classes/Main/I18nManager.swift:34]:
```suggestion
    ///   language, even if an appropriate translation exists in the preferred language, this can lead to unintended substitutions. 
    /// To avoid having the issue in the app, use more descriptive translation keys or use case-sensitive words as a translated word (e.g., "ok" != "Ok", "done" != "Done", etc.).
```
